File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int start = 0, end = 0, maxLength = 0;
        while(end < nums.length){
            if(nums[end] == 1){
                end++;
                maxLength = Math.max(end-start, maxLength);
            }
            else if(nums[end] == 0 && k > 0){
                end++;
                k--;
                maxLength = Math.max(end-start, maxLength);
            }else{
                if(nums[start] == 0) k++;
                start++;
            }
        }
        return maxLength;
    }
}

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        start = 0
        res = 0
        
        for end in range(len(nums)):
            if nums[end] == 0:
                k -= 1
            
            while k < 0:
                if nums[start] == 0:
                    k += 1
                start += 1
            
            res = max(res, end - start + 1)
        
        return res